### PR TITLE
ARROW-11240: [Packaging][R] Add mimalloc to R packaging

### PR DESCRIPTION
**Pull request by ianmcook**

### CODE REVIEW

File: r/README.md [hunk 1]
```diff
@@ -103,6 +103,7 @@ cmake
   -DARROW_DATASET=ON \
   -DARROW_FILESYSTEM=ON \
   -DARROW_JEMALLOC=ON \
+  -DARROW_MIMALLOC=ON \
```
Comment (nealrichardson) [inline r/README.md:106]: Maybe list this as optional below

**Comment by nealrichardson** [comment]
@github-actions crossbow submit -g r

**Comment by nealrichardson** [comment]
We have two problems here:

1) mimalloc appears not to be supported on gcc 4.8
2) the devtoolset-8 CI job that @jonkeane added appears to be using gcc 4.8 now

**Comment by jonkeane** [comment]
I'm looking into the devtoolset-8 issue, and sure enough at _one point_ in that PR it was successful in using gcc 8.3, but the final runs / commits that were merged appear to revert back to using 4.8. It looks like the change that borked things was https://github.com/apache/arrow/pull/8932/commits/4643401d43b880789962e47e50e0576e341437ef. You could try reverting that to see if the devtools-8 crossbow builds are using gcc 8.3 without it.

**Comment by nealrichardson** [comment]
👍 that can be done on a separate issue, don't need to block this. gcc 4.8 workaround is still needed for the other build.

**Comment by ianmcook** [comment]
Do we know what minimum GCC version mimalloc requires? I dug around in the mimalloc repo and various other places but couldn't find any indication of that

File: r/tools/linuxlibs.R [hunk 2]
```diff
@@ -412,6 +413,23 @@ with_s3_support <- function(env_vars) {
...
 }
 
+with_mimalloc <- function(env_vars) {
+  arrow_mimalloc <- toupper(Sys.getenv("ARROW_MIMALLOC")) == "ON" || tolower(Sys.getenv("LIBARROW_MINIMAL")) == "false"
+  if (arrow_mimalloc) {
+  # User wants mimalloc. Let's make sure they're not on gcc < 4.9
+    info <- system(paste("export", env_vars, "&& $CMAKE --system-information"), intern = TRUE)
+    info <- grep("^[A-Z_]* .*$", info, value = TRUE)
+    vals <- as.list(sub('^.*? "?(.*?)"?$', "\\1", info))
+    names(vals) <- sub("^(.*?) .*$", "\\1", info)
```
Comment (ianmcook) [inline r/tools/linuxlibs.R:423]: This is copy-pasted from `with_s3_support()` above. Consider factoring out into a separate function

**Comment by ianmcook** [comment]
@github-actions crossbow submit test-r-rstudio-r-base-3.6-*

File: r/tools/linuxlibs.R (hunk 2 above)
Comment (nealrichardson) [reply r/tools/linuxlibs.R:423]: Yes, let's factor out the gcc 4.9 check to a function

**Comment by nealrichardson** [comment]
> Do we know what minimum GCC version mimalloc requires?

My guess is that it needs full C++11 support, and so that just means gcc > 4.8, so I think this is fine.

File: r/README.md (hunk 1 above)
Comment (ianmcook) [reply r/README.md]: Done in a59b9a2. I noticed the S3 flag wasn't in the list of optional flags here so I added that as well

**Comment by pitrou** [comment]
mimalloc is pure C, but it probably needs some system calls that are not present on old glibcs or kernels.

**Comment by nealrichardson** [comment]
Ah my bad. It's specifically a gcc issue fixed in 4.9: https://stackoverflow.com/questions/20326604/stdatomic-h-in-gcc-4-8

File: r/tools/linuxlibs.R (hunk 2 above)
Comment (ianmcook) [reply r/tools/linuxlibs.R]: Factored out in 6b5a193

**Comment by ianmcook** [comment]
The crossbow jobs that were failing because of gcc 4.8 are now succeeding after the latest commits. The install logs in both jobs show the following
```
**** mimalloc support not available for gcc < 4.9; building with ARROW_MIMALLOC=OFF
```

**Comment by nealrichardson** [comment]
Great, can you (1) remove [WIP] from the issue title and then (2) resolve the merge conflict, which will also trigger CI.

**Review by nealrichardson — APPROVED** [review]

**Pull request closed without merging**